### ISSUE TITLE
CRAN release & adonis deprecation: Request for Comments

### ISSUE THREAD

**Issue by jarioksa** [vegandevs/vegan#523]
I am planning a new CRAN release of **vegan** in the latter half of August. There are quite a few changes, although nothing very dramatic. For most users the main thing may be wider and robuster support for **tidy** dialect and **ggplot** graphics. See [NEWS](https://github.com/vegandevs/vegan/blob/master/inst/NEWS.md) for details of the numerous other changes.

For release the main thing may be that I had planned to deprecate `adonis` in favour of `adonis2` in this release. `adonis2` was released in **vegan** 2.4-0 on Jun-15 2016 or more than six years ago, but it seems that most dependent packages still use `adonis`. I ran reverse dependence tests, and found out that six packages will issue a WARNING if `adonis` is deprecated: **betaper**, **GUniFrac**, **microbial**, **RVAideMemoire**, **SigTree** and **smartsnp**. In addition, **QFASA** passes tests without WARNING, but uses `adonis` without explicit import as `vegan::adonis()`. Similar test is performed in CRAN and new WARNINGs are flagged as changes to "worse", and may cause ripples in the smooth submission process. Therefore I'd like to avoid those changes to "worse". I planned to deprecate `adonis` already in the current release 2.6-2, but just to avoid problems I changed `warning()` to `message("'adonis' will be deprecated: use 'adonis2' instead")` which is not flagged in the CRAN process. I also emailed to those package authors on 15-Apr 2022. My dependence tests were less complete then, and did not include many packages dependent on BioConductor, and I only found five "worse" cases. Two of these were fixed promptly (**dispRity** and **microeco**), but three are still in the current list of regressions.

I'm still uncertain what to do or what should be the deprecation policy. My initial plan was just to phase out and remove `adonis` and leave only `adonis2`. However, **dispRity** defines an S3 method for `adonis` (and correctly handles `adonis2`), and obviously prefers to keep the name. One choice is to keep the name `adonis` in **vegan**, but eventually synonymize it with `adonis2`. The API is similar in both and if the function is just used via **vegan** API, users should see no difference (`adonis2` offers more choices, but default is similar as old `adonis`). However, if packages try to access the internal structures of the result object, they will fail because intestines of `adonis` and `adonis2` are completely different. 

One way to proceed is:
1. Keep the 2.6-2 behaviour in the next release: print `message()` of future deprecation, but run `adonis`.
2. In a future release, print `message`, but run `adonis2`. That is, to remove `adonis` code and make commands `adonis` and `adonis2` synonyms.  This will give an error if there is a package that tries to access internals of the result object, but normal use should see no difference.

I just wonder, if there should be one release in between with `warning` to alert those package authors that access the `adonis2` result object directly.

CC: @cstewartGH @ChristianHuber @TGuillerme @ChiLiubio

**Comment by TGuillerme** [comment]
Thanks for the heads up @jarioksa! I'm happy to go with either option (i.e. the release in between with `warning` or the straight synonymy). I'm also happy to change the `S3` in `dispRity` to `.adonis2` if that helps (I kept it as `.adonis` since I imagine people are familiar with that function name - but I might be wrong). I'm going to publish a release (1.7) this week so I can quickly add the change before it goes to CRAN.

**Comment by gavinsimpson** [comment]
I've done a PR on QFASA to add an `importFrom(vegan, adonis)` statement: cstewartGH/QFASA#13 But I assume this isn't a major problem other than the fact that it still uses `adonis()` and @jarioksa you wanted to throw deprecation messages that would complicated getting this on CRAN?

I think a recent commit (23f683329772f2d26e6534728514113be1ad6bb5) from you @jarioksa has loosened the deprecation warnings to just a message. So I assume you are going with something like your wondering "if there should be one release in between with warning to alert those package authors that access the `adonis` result object directly."

My personal opinion would be that we eventually make `adonis()` defunct (so we must deprecate it first) - I want old code to error rather than silently (or otherwise) return a different object if the version of vegan I am using has changed that much.

I appreciate this will cause some CRAN headaches but so long as we follow the approved route of contacting maintainers, then deprecating `adonis()` officially for a point release before making it defunct in a later point release, the problem will become that of other package maintainers to either update their code to use `adonis2()` or find some other solution.

**Comment by cstewartGH** [comment]
Thank you for this!  I have updated the QFASA package with adonis2() (which required a few changes to my code) but I am still in the process of making other changes to the package and haven't submitted it to CRAN yet.

**Comment by jarioksa** [comment]
**vegan** 2.6-4 is now on CRAN. It will extend the mercy period and only print a message of future deprecation. That deprecation will happen when the current devel version (2.6-5) will be released. The following packages still use `adonis` according to reverse dependency tests: **betaper**, **GUniFrac**, **microbial**, **RVAideMemoire**, **SigTree**, **smartsnp**.

**Comment by christianangulo450** [comment]
Hello, dear friends. I would like to know how I can get the pdf. file to learn how to use vegan.

**Comment by jarioksa** [comment]
`adonis` was deprecated in CRAN release 2.6-6 and is defunct in github version 2.7-0. This breaks some CRAN packages. Issue #647 monitors the breaking packages, and Announcement #641 gives instructions of adapting your packages and scripts to the changes.

**Comment by cstewartGH** [comment]
Thank you for the notification!  I had previously changed adonis to adonis2 in a past QFASA package update. 😊

Connie

CONNIE STEWART (She/Her/Hers)<http://www.unb.ca/pronouns>
On Sabbatical Leave
Professor and Department Chair • Mathematics and Statistics, Department of
T 506 648-5725
[University of New Brunswick]
[Facebook]/uofnb<https://www.facebook.com/uofnb> [Twitter] @unb<https://twitter.com/UNB> [Instagram]  @discoverunb<httpss://instagram.com/discoverunb/> UNB.ca<https://www.unb.ca/>
UNB stands on the unsurrendered and unceded traditional Wolastoqey land as established in a series of Peace and Friendship Treaties.
Confidentiality Note: This email and the information contained in it is confidential, may be privileged and is intended for the exclusive use of the addressee(s). Any other person is strictly prohibited from using, disclosing, distributing or reproducing it. If you have received this communication in error, please reply by email to the sender and delete or destroy all copies of this message.

**Comment by jarioksa** [comment]
@cstewartGH , yes I know your package is OK. I was closing some old and solved issues, and github automatically sent messages to people involved in those issues. This was an unfortunate flashback.  My apologies.

**Comment by cstewartGH** [comment]
That’s great!  I appreciate the confirmation that I hadn’t missed something. 😊

Connie

**Comment by jarioksa** [comment]
`adonis` was formally defunct in **vegan** release 2.7-1 (on CRAN on 2025-06-05). Use `adonis2`.
